File: worker/src/main/java/com/github/k0kubun/github_ranking/worker/RepositoryRankingWorker.java
package com.github.k0kubun.github_ranking.worker;

import com.github.k0kubun.github_ranking.config.Config;
import com.github.k0kubun.github_ranking.dao.repository.RepositoryDao;
import com.github.k0kubun.github_ranking.dao.repository.RepositoryRankDao;
import com.github.k0kubun.github_ranking.model.Repository;
import com.github.k0kubun.github_ranking.model.RepositoryRank;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

import org.skife.jdbi.v2.DBI;
import org.skife.jdbi.v2.Handle;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RepositoryRankingWorker
        extends Worker
{
    private static final int ITERATE_MIN_STARS = 10;
    private static final Logger LOG = LoggerFactory.getLogger(RepositoryRankingWorker.class);

    private final BlockingQueue<Boolean> repoRankingQueue;
    private final DBI dbi;

    public RepositoryRankingWorker(Config config)
    {
        super();
        repoRankingQueue = config.getQueueConfig().getRepoRankingQueue();
        dbi = new DBI(config.getDatabaseConfig().getDataSource());
    }

    @Override
    public void perform()
            throws Exception
    {
        while (repoRankingQueue.poll(5, TimeUnit.SECONDS) == null) {
            if (isStopped) {
                return;
            }
        }
        LOG.info("----- started RepositoryRankingWorker -----");
        try (Handle handle = dbi.open()) {
            RepositoryRank lastRank = updateUpperRanking(handle);
            if (lastRank != null) {
                updateLowerRanking(handle, lastRank);
            }
        }
        LOG.info("----- finished RepositoryRankingWorker -----");
    }

    private RepositoryRank updateUpperRanking(Handle handle)
    {
        int count = handle.attach(RepositoryDao.class).countRepos(); // warmup
        PaginatedRepositories paginatedRepos = new PaginatedRepositories(handle);
        List<Repository> repos;

        List<RepositoryRank> commitPendingRanks = new ArrayList<>(); // listed in stargazers_count DESC
        RepositoryRank currentRank = null;
        int currentRankNum = 0;

        while (!(repos = paginatedRepos.nextRepos()).isEmpty()) {
            // Shutdown immediately if requested, even if it's in progress.
            if (isStopped) {
                return null;
            }

            for (Repository repo : repos) {
                if (currentRank == null) {
                    currentRank = new RepositoryRank(repo.getStargazersCount(), 1);
                    currentRankNum = 1;
                }
                else if (currentRank.getStargazersCount() == repo.getStargazersCount()) {
                    currentRankNum++;
                }
                else {
                    commitPendingRanks.add(currentRank);
                    currentRank = new RepositoryRank(repo.getStargazersCount(), currentRank.getRank() + currentRankNum);
                    currentRankNum = 1;
                }
            }

            if (!commitPendingRanks.isEmpty()) {
                commitRanks(handle, commitPendingRanks);
                commitPendingRanks.clear();
            }
            int rows = currentRank.getRank() + currentRankNum - 1;
            LOG.info("RepositoryRankingWorker (" + calcProgress(rows, count) + ", " + Integer.valueOf(rows).toString() +
                    "/" + Integer.valueOf(count).toString() + " rows, rank " + Integer.valueOf(currentRank.getRank()).toString() + ", " +
                    Integer.valueOf(currentRank.getStargazersCount()).toString() + " stars)");

            // Switch the way to calculate ranking under 10 stars
            if (currentRank.getStargazersCount() <= ITERATE_MIN_STARS) {
                return currentRank;
            }
        }
        return currentRank;
    }

    private void updateLowerRanking(Handle handle, RepositoryRank lastRepoRank)
    {
        List<RepositoryRank> repoRanks = new ArrayList<>(); // listed in stargazers_count DESC
        repoRanks.add(lastRepoRank);

        int lastRank = lastRepoRank.getRank();
        for (int lastStars = lastRepoRank.getStargazersCount(); lastStars > 0; lastStars--) {
            LOG.info("RepositoryRankingWorker for " + Integer.valueOf(lastStars - 1).toString());
            int count = handle.attach(RepositoryDao.class).countReposHavingStars(lastStars);
            repoRanks.add(new RepositoryRank(lastStars - 1, lastRank + count));
            lastRank += count;
        }
        commitRanks(handle, repoRanks);
    }

    private void commitRanks(Handle handle, List<RepositoryRank> repoRanks)
    {
        // `repoRanks` is listed in stargazers_count DESC
        Integer minStars = lastOf(repoRanks).getStargazersCount();
        Integer maxStars = repoRanks.get(0).getStargazersCount();

        handle.useTransaction((conn, status) -> {
            conn.attach(RepositoryRankDao.class).deleteBetween(minStars, maxStars);
            conn.attach(RepositoryRankDao.class).bulkInsert(repoRanks);
        });
    }

    private RepositoryRank lastOf(List<RepositoryRank> repoRanks)
    {
        return repoRanks.get(repoRanks.size() - 1);
    }

    private String calcProgress(int child, int parent)
    {
        return String.format("%.3f%%", (float) child / (float) parent);
    }

    // This class does cursor-based-pagination for repositories order by stargazers_count DESC.
    public class PaginatedRepositories
    {
        private static final int PAGE_SIZE = 5000;

        private final RepositoryDao repoDao;
        private Integer lastMinStars;
        private Long lastMinId;

        public PaginatedRepositories(Handle handle)
        {
            repoDao = handle.attach(RepositoryDao.class);
            lastMinStars = null;
            lastMinId = null;
        }

        public List<Repository> nextRepos()
        {
            List<Repository> repos;
            if (lastMinId == null && lastMinStars == null) {
                repos = repoDao.starsDescFirstRepos(PAGE_SIZE);
            }
            else {
                repos = repoDao.starsDescReposAfter(lastMinStars, lastMinId, PAGE_SIZE);
            }
            if (repos.isEmpty()) {
                return repos;
            }

            Repository lastRepo = repos.get(repos.size() - 1);
            lastMinStars = lastRepo.getStargazersCount();
            lastMinId = lastRepo.getId();
            return repos;
        }
    }
}

File: worker/src/main/java/com/github/k0kubun/gitstar_ranking/worker/RepositoryRankingWorker.kt
package com.github.k0kubun.gitstar_ranking.worker

import com.github.k0kubun.gitstar_ranking.config.Config
import com.github.k0kubun.gitstar_ranking.model.Repository
import java.util.concurrent.BlockingQueue
import org.skife.jdbi.v2.DBI
import kotlin.Throws
import java.lang.Exception
import com.github.k0kubun.gitstar_ranking.model.RepositoryRank
import com.github.k0kubun.gitstar_ranking.repository.dao.RepositoryDao
import org.skife.jdbi.v2.TransactionStatus
import com.github.k0kubun.gitstar_ranking.repository.dao.RepositoryRankDao
import java.util.ArrayList
import org.skife.jdbi.v2.Handle
import org.slf4j.LoggerFactory

private const val PAGE_SIZE = 5000

class RepositoryRankingWorker(config: Config) : Worker() {
    private val repoRankingQueue: BlockingQueue<Boolean>
    private val dbi: DBI
    @Throws(Exception::class)
    override fun perform() {
        LOG.info("----- started RepositoryRankingWorker -----")
        dbi.open().use { handle ->
            val lastRank = updateUpperRanking(handle)
            lastRank?.let { updateLowerRanking(handle, it) }
        }
        LOG.info("----- finished RepositoryRankingWorker -----")
    }

    private fun updateUpperRanking(handle: Handle): RepositoryRank? {
        val count = handle.attach(RepositoryDao::class.java).countRepos() // warmup
        val paginatedRepos = PaginatedRepositories(handle)
        var repos: List<Repository>
        val commitPendingRanks: MutableList<RepositoryRank> = ArrayList() // listed in stargazers_count DESC
        var currentRank: RepositoryRank? = null
        var currentRankNum = 0
        while (!paginatedRepos.nextRepos().also { repos = it }.isEmpty()) {
            // Shutdown immediately if requested, even if it's in progress.
            if (isStopped) {
                return null
            }
            for (repo in repos) {
                if (currentRank == null) {
                    currentRank = RepositoryRank(repo.stargazersCount, 1)
                    currentRankNum = 1
                } else if (currentRank.stargazersCount == repo.stargazersCount) {
                    currentRankNum++
                } else {
                    commitPendingRanks.add(currentRank)
                    currentRank = RepositoryRank(repo.stargazersCount, currentRank.rank + currentRankNum)
                    currentRankNum = 1
                }
            }
            if (!commitPendingRanks.isEmpty()) {
                commitRanks(handle, commitPendingRanks)
                commitPendingRanks.clear()
            }
            val rows = currentRank!!.rank + currentRankNum - 1
            LOG.info("RepositoryRankingWorker (" + calcProgress(rows, count) + ", " + Integer.valueOf(rows).toString() +
                "/" + Integer.valueOf(count).toString() + " rows, rank " + Integer.valueOf(currentRank.rank).toString() + ", " +
                Integer.valueOf(currentRank.stargazersCount).toString() + " stars)")

            // Switch the way to calculate ranking under 10 stars
            if (currentRank.stargazersCount <= ITERATE_MIN_STARS) {
                return currentRank
            }
        }
        return currentRank
    }

    private fun updateLowerRanking(handle: Handle, lastRepoRank: RepositoryRank) {
        val repoRanks: MutableList<RepositoryRank> = ArrayList() // listed in stargazers_count DESC
        repoRanks.add(lastRepoRank)
        var lastRank = lastRepoRank.rank
        for (lastStars in lastRepoRank.stargazersCount downTo 1) {
            LOG.info("RepositoryRankingWorker for " + Integer.valueOf(lastStars - 1).toString())
            val count = handle.attach(RepositoryDao::class.java).countReposHavingStars(lastStars)
            repoRanks.add(RepositoryRank(lastStars - 1, lastRank + count))
            lastRank += count
        }
        commitRanks(handle, repoRanks)
    }

    private fun commitRanks(handle: Handle, repoRanks: List<RepositoryRank>) {
        // `repoRanks` is listed in stargazers_count DESC
        val maxStars = repoRanks[0].stargazersCount
        val highestRank = repoRanks[0].rank
        val minStars = lastOf(repoRanks).stargazersCount
        val lowestRank = lastOf(repoRanks).rank
        handle.useTransaction { conn: Handle, status: TransactionStatus? ->
            conn.attach(RepositoryRankDao::class.java).deleteStarsBetween(minStars, maxStars)
            conn.attach(RepositoryRankDao::class.java).deleteRankBetween(highestRank, lowestRank)
            conn.attach(RepositoryRankDao::class.java).bulkInsert(repoRanks)
        }
    }

    private fun lastOf(repoRanks: List<RepositoryRank>): RepositoryRank {
        return repoRanks[repoRanks.size - 1]
    }

    private fun calcProgress(child: Int, parent: Int): String {
        return String.format("%.3f%%", child.toFloat() / parent.toFloat())
    }

    // This class does cursor-based-pagination for repositories order by stargazers_count DESC.
    inner class PaginatedRepositories(handle: Handle) {
        private val repoDao: RepositoryDao
        private var lastMinStars: Int?
        private var lastMinId: Long?
        fun nextRepos(): List<Repository> {
            val repos: List<Repository>
            repos = if (lastMinId == null && lastMinStars == null) {
                repoDao.starsDescFirstRepos(PAGE_SIZE)
            } else {
                repoDao.starsDescReposAfter(lastMinStars, lastMinId, PAGE_SIZE)
            }
            if (repos.isEmpty()) {
                return repos
            }
            val lastRepo = repos[repos.size - 1]
            lastMinStars = lastRepo.stargazersCount
            lastMinId = lastRepo.id
            return repos
        }

        init {
            repoDao = handle.attach(RepositoryDao::class.java)
            lastMinStars = null
            lastMinId = null
        }
    }

    companion object {
        private const val ITERATE_MIN_STARS = 10
        private val LOG = LoggerFactory.getLogger(RepositoryRankingWorker::class.java)
    }

    init {
        repoRankingQueue = config.queueConfig.repoRankingQueue
        dbi = DBI(config.databaseConfig.dataSource)
    }
}
